Initial implementation of slider widget.

raghucssit opened the pull request.

DieterMai commented:
Your implementation looks way different then what see for native windows. Did you use some other native implementation as reference?
![slider compare](https://github.com/user-attachments/assets/16a029b5-94dc-4852-a8b5-d1b4754bc166)

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
```
You may add yourself.

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+			case SWT.MouseUp -> onMouseUp(event);
+			case SWT.MouseHorizontalWheel -> onMouseHorizontalWheel(event);
+			case SWT.MouseVerticalWheel -> onMouseVerticalWheel(event);
+			case SWT.Paint -> onPaint(event);
```
I'd order the events by importance (most important Paint should come first), but otherwise good to use the light-weight events.

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+		orientation = style & (SWT.LEFT_TO_RIGHT | SWT.RIGHT_TO_LEFT);
+
+		super.style |= horizontal ? SWT.HORIZONTAL : SWT.VERTICAL;
+
```
Please remove empty lines after { or before }.

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+	private void onMouseVerticalWheel(Event event) {
+		if (!isVisible()) {
+			return;
+		}
```
IMHO, if not visible, there should not be any event at all. Did you got some while invisible?

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+
+	private void onMouseHorizontalWheel(Event event) {
+		if (!isVisible()) {
+			return;
```
same

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+		}
+
+		if (event.count > 0) {
+			increment(2);
```
What means this hard-coded `2`? Why not `1`?

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+	private void onPaint(Event event) {
+		if (!isVisible()) {
+			return;
+		}
```
SWT.Paint should not be sent if not visible.

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+	}
+
+	private void onMouseDown(Event event) {
+		// Drag of the thumb.
```
Should it work for any mouse button? If not, please check with `if (event.button != 1) return;`

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+
+
+	private void onMouseUp(Event event) {
+		if (isDragging) {
```
Try with moving the mouse outside the shell and moving it back. There exists a method to get all mouse events (forgot its name).

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+	}
+
+	private void onMouseMove(Event event) {
+
```
Remove empty line after {

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+		} else {
+			thumbHovered = false;
+			redraw();
+		}
```
The if-statement is not necessary - `thumbHovered = isThumbHovered; redraw();` should do the same.

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+		int height = drawingArea.height;
+
+		// Fill background
+		if (getBackground() != null) {
```
getBackground() will always return != null.

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+
+		// Fill background
+		if (getBackground() != null) {
+			gc.setBackground(getBackground());
```
Don't invoke the getter twice but store in local variable.

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+	public void removeSelectionListener(SelectionListener listener) {
+		checkWidget();
+		if (listener == null)
+			error(SWT.ERROR_NULL_ARGUMENT);
```
Should be on same line or as block statement (`{ ... }`).

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+	 *                         </ul>
+	 */
+	public int getIncrement() {
+		checkWidget();
```
Probably these checks can be removed for such plain getters. Later, the separate Renderer needs to use these methods and checking for each method is useless wasted CPU time.

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+	private int minMax(int min, int value, int max) {
+		return Math.min(Math.max(min, value), max);
+	}
+
```
remove empty line

tmssngr commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+		this.pageIncrement = pageIncrement;
+	}
+
+	private int minMax(int min, int value, int max) {
```
Should be `static`. Maybe it could be moved to a different utilities class?

raghucssit commented:
> Your implementation looks way different then what see for native windows. Did you use some other native implementation as reference? ![slider compare](https://private-user-images.githubusercontent.com/26310735/423164666-16a029b5-94dc-4852-a8b5-d1b4754bc166.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyMTA3ODAsIm5iZiI6MTc0MjIxMDQ4MCwicGF0aCI6Ii8yNjMxMDczNS80MjMxNjQ2NjYtMTZhMDI5YjUtOTRkYy00ODUyLWE4YjUtZDFiNDc1NGJjMTY2LlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE3VDExMjEyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMxNWJkZDYyZThiOWYxNWYxNTQ4YzQ5YWIxNGExMWIxNjI1ZWU4NzRmZWRjM2E1YzA1OTI2MWEwZDJjMDZiZGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.23KBM9JiXyJ6-Jb7aEmhcbl_lJKBnFhPSTaoRZOZINY)

My implementation is similar to native Slider on RHEL-9.2. Can you please share your environment details like OS and Style settings.
Below is my testing code:
```
        Slider slider = new Slider(deftSlidComp, SWT.HORIZONTAL);
        slider.setMaximum(25);
        slider.setMinimum(0);
        slider.setIncrement(2);
        slider.setPageIncrement(5);
        slider.setThumb(5);
        slider.setSelection(0);
        slider.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));
```

Example: Left one is Native and Right one is SkiJa drawm.

https://github.com/user-attachments/assets/733863e1-f23e-4cba-a118-f6b62c5ba4bc

raghucssit commented:
> Could you please add a simple snippet to test Slider?

This is the complete example. Slider widget is present twice because i was having both native and new drawn next to each other before i rename new and delete the old one. I was setting same attributes for each and drawing similar to native..

```
import org.eclipse.swt.SWT;
import org.eclipse.swt.graphics.Color;
import org.eclipse.swt.layout.GridData;
import org.eclipse.swt.layout.GridLayout;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.Event;
import org.eclipse.swt.widgets.Label;
import org.eclipse.swt.widgets.Listener;
import org.eclipse.swt.widgets.Shell;
import org.eclipse.swt.widgets.Slider;

public class SliderTest {


    public static void main(String[] args) {
        Display display = new Display();
        Shell shell = new Shell(display);
        shell.setText("SWT Slider Example");
        shell.setSize(800, 600);
        
        Color colorCyan = display.getSystemColor(SWT.COLOR_CYAN);
        Color colorWhite = display.getSystemColor(SWT.COLOR_WHITE);
        
        // Set GridLayout with margins for Shell
        GridLayout shellLayout = new GridLayout(1, false);
        shellLayout.marginWidth = 30;
        shellLayout.marginHeight = 30;
        shellLayout.horizontalSpacing = 10;
        shellLayout.verticalSpacing = 10;
        shell.setLayout(shellLayout);
        
        GridData gd = new GridData(SWT.FILL, SWT.FILL, true, true);
        
        Composite deftSlidComp = new Composite(shell, SWT.BORDER);
        deftSlidComp.setLayout(shellLayout);
        deftSlidComp.setLayoutData(gd);
        
        Label title = new Label(deftSlidComp, SWT.BORDER);
        title.setText("Default Slider");
        title.setLayoutData(gd);
        
        Slider slider = new Slider(deftSlidComp, SWT.HORIZONTAL);
        slider.setMaximum(25);
        slider.setMinimum(0);
        slider.setIncrement(2);
        slider.setPageIncrement(5);
        slider.setThumb(5);
        slider.setSelection(0);
        slider.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));
//        slider.setEnabled(false);
        
//        slider.setBackground(colorCyan);

        Slider custSlider = new Slider(deftSlidComp, SWT.HORIZONTAL);
        custSlider.setMaximum(25);
        custSlider.setMinimum(0);
        custSlider.setIncrement(2);
        custSlider.setPageIncrement(5);
        custSlider.setThumb(5);
        custSlider.setSelection(0);
        custSlider.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));
//        custSlider.setEnabled(false);
        
//        custSlider.setBackground(colorCyan);
        
        //custom slider
        Composite custtSlidComp = new Composite(shell, SWT.BORDER);
        custtSlidComp.setLayout(shellLayout);
        custtSlidComp.setLayoutData(gd);
        
        Label custtitle = new Label(custtSlidComp, SWT.BORDER);
        custtitle.setText("Default Slider");
        custtitle.setLayoutData(gd);
        
        
        Label value = new Label(deftSlidComp, SWT.BORDER);
        String string = "Selected Value: ";
		value.setText(string);
        value.setLayoutData(gd);
        
        slider.addListener(SWT.Selection, new Listener() {
            public void handleEvent(Event event) {
                int perspectiveValue = slider.getMaximum() - slider.getSelection() + slider.getMinimum() - slider.getThumb();
                value.setText(string + slider.getSelection());
              }
            });
  
        
        
        Label custSlidValue = new Label(custtSlidComp, SWT.BORDER);
		custSlidValue.setText(string);
        custSlidValue.setLayoutData(gd);
        
        custSlider.addListener(SWT.Selection, new Listener() {
            public void handleEvent(Event event) {
                custSlidValue.setText(string + custSlider.getSelection());
              }
            });
        
        
        
//        shell.pack();
        shell.open();
        while (!shell.isDisposed()) {
            if (!display.readAndDispatch()) {
                display.sleep();
            }
        }
        colorCyan.dispose();
        colorWhite.dispose();
        display.dispose();
    }



}

```

raghucssit commented:
> > > Your implementation looks way different then what see for native windows. Did you use some other native implementation as reference?
> > 
> > 
> > My implementation is similar to native Slider on RHEL-9.2. Can you please share your environment details like OS and Style settings.
> 
> I use Windows
> 
> > Below is my testing code:
> > ```
> >         Slider slider = new Slider(deftSlidComp, SWT.HORIZONTAL);
> >         slider.setMaximum(25);
> >         slider.setMinimum(0);
> >         slider.setIncrement(2);
> >         slider.setPageIncrement(5);
> >         slider.setThumb(5);
> >         slider.setSelection(0);
> >         slider.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));
> > ```
> 
> I would recommend using the control example. I saw some differences there too.

I can confirm your findings on Windows 10. But i don't know what we can do about it now.. I think we are trying to implement something which looks same on all platforms. Otherwise we have to add check and draw it differently.
![image](https://github.com/user-attachments/assets/6eaaff20-a9fb-4291-8447-21d786054d7e)

raghucssit commented:
> I meant, please add it to your branch, e.g. below `tests/org.eclipse.swt.tests/ManualTests/`.
> 
> On Windows 11 it looks like this (hovering makes it looks weird): ![screencast](https://private-user-images.githubusercontent.com/6029561/423495211-abe83f0d-0cf8-4ab6-bd04-4187ddff2430.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyMTg4MzcsIm5iZiI6MTc0MjIxODUzNywicGF0aCI6Ii82MDI5NTYxLzQyMzQ5NTIxMS1hYmU4M2YwZC0wY2Y4LTRhYjYtYmQwNC00MTg3ZGRmZjI0MzAuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMTdUMTMzNTM3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWE0ZDk1NzI2MDU4OWU1YmYxZGMyYjY2YjM4OWVhOGE2MWQ1OGI1OTkzYjcyZTM0NzYzMjExMWU0YzljNzcxMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.rz09n7Ymsz27vmZDZd3z2ZZOv5hRUX04B8FAIkhXqjY) [ ![screencast](https://private-user-images.githubusercontent.com/6029561/423495211-abe83f0d-0cf8-4ab6-bd04-4187ddff2430.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyMTg4MzcsIm5iZiI6MTc0MjIxODUzNywicGF0aCI6Ii82MDI5NTYxLzQyMzQ5NTIxMS1hYmU4M2YwZC0wY2Y4LTRhYjYtYmQwNC00MTg3ZGRmZjI0MzAuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMTdUMTMzNTM3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWE0ZDk1NzI2MDU4OWU1YmYxZGMyYjY2YjM4OWVhOGE2MWQ1OGI1OTkzYjcyZTM0NzYzMjExMWU0YzljNzcxMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.rz09n7Ymsz27vmZDZd3z2ZZOv5hRUX04B8FAIkhXqjY) ](https://private-user-images.githubusercontent.com/6029561/423495211-abe83f0d-0cf8-4ab6-bd04-4187ddff2430.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyMTg4MzcsIm5iZiI6MTc0MjIxODUzNywicGF0aCI6Ii82MDI5NTYxLzQyMzQ5NTIxMS1hYmU4M2YwZC0wY2Y4LTRhYjYtYmQwNC00MTg3ZGRmZjI0MzAuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMTdUMTMzNTM3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWE0ZDk1NzI2MDU4OWU1YmYxZGMyYjY2YjM4OWVhOGE2MWQ1OGI1OTkzYjcyZTM0NzYzMjExMWU0YzljNzcxMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.rz09n7Ymsz27vmZDZd3z2ZZOv5hRUX04B8FAIkhXqjY) [ ](https://private-user-images.githubusercontent.com/6029561/423495211-abe83f0d-0cf8-4ab6-bd04-4187ddff2430.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyMTg4MzcsIm5iZiI6MTc0MjIxODUzNywicGF0aCI6Ii82MDI5NTYxLzQyMzQ5NTIxMS1hYmU4M2YwZC0wY2Y4LTRhYjYtYmQwNC00MTg3ZGRmZjI0MzAuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMTdUMTMzNTM3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWE0ZDk1NzI2MDU4OWU1YmYxZGMyYjY2YjM4OWVhOGE2MWQ1OGI1OTkzYjcyZTM0NzYzMjExMWU0YzljNzcxMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.rz09n7Ymsz27vmZDZd3z2ZZOv5hRUX04B8FAIkhXqjY)
> 
> It would be helpful if the snippet would also contain a vertical slider, too.

Yeah. I too felt it. It is the background color of track which makes it look differ. Even the thumb color does not match 100%.
I have picked the color which best fits from color picker. May be i will try to refine the color. Please provide a patch if you find exact matching RGB.

raghucssit commented:
> Could you please also extract a new SliderRenderer (subclass from ControlRenderer) and either LinuxSliderRenderer or BasicSliderRender with your implementation? Take a look at, e.g., Button how it should be done. Thanks in advance.

Yes. I will fix review comments along with extracting it as a SliderRenderer.

tmssngr commented:
> It is the background color of track which makes it look differ. Even the thumb color does not match 100%. I have picked the color which best fits from color picker. May be i will try to refine the color. Please provide a patch if you find exact matching RGB.

No need to overcomplicate the matter. The color fine tuning can happen later.

DieterMai commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+			return;
+		}
+
+		redraw();
```
You redraw when minimum has not changed, but you need to redraw when it has changed

DieterMai commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+		// Ensure selection does not exceed max - thumb
+		int newValue = Math.max(min, Math.min(selection, max - thumb));
+
+		selectAndNotify(newValue);
```
redraw is missing

DieterMai commented on bundles/org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/widgets/Slider.java:
```diff
@@ -0,0 +1,903 @@
+
+		if (thumb <= this.maximum && thumb <= this.selection) {
+			this.thumb = thumb;
+			return;
```
return before triggering redraw

DieterMai reviewed (commented):
Some redraws are missing

raghucssit commented:
@tmssngr , @DieterMai 
Thank you for your review. I have fixed all the mentioned comments.
Also I have refactored it as a SliderRenderer.
Please check and merge it.

DieterMai commented:
I see all the event handling is in the renderer now. In my mind the renderer should only handle the rendering and the widget should take care of the event handling.

tmssngr commented:
> I see all the event handling is in the renderer now. In my mind the renderer should only handle the rendering and the widget should take care of the event handling.

I agree with Dieter.

raghucssit commented:
@tmssngr @DieterMai 
I have addressed the design issue. I moved event handling to renderer by assuming any custom implementation may need custom event handling. Unfortunately it was going against the design of renderer. Now I have fixed it.

tmssngr commented:
I have added some fixup commits. If you agree with these changes, please squash all commits in your branch and *Rebase and merge* it.

Overall good work! Thank you.

raghucssit commented:
> I have added some fixup commits. If you agree with these changes, please squash all commits in your branch and _Rebase and merge_ it.
> 
> Overall good work! Thank you.

Shall I merge then ?
